src/CowAmmModule.sol

```diff
@@ -0,0 +1,358 @@
+import {IConditionalOrder} from "lib/composable-cow/src/BaseConditionalOrder.sol";
+
+/**
+ * @title CoW AMM Module Factory
+ * @author CoW Protocol Developers
+ * @dev A factory for creating CoW AMM modules.
+ */
+contract CowAmmModuleFactory {
+    using Clones for address;
+
+    /**
+     * @notice The implementation of the CoW AMM module with the respective network-specific addresses and amm handler.
+     */
+    CowAmmModule public immutable IMPLEMENTATION;
+
+    /**
+     * @dev Error messages for the factory.
+     */
+    error InvalidParameter();
+
+    /**
+     * @notice Index all cow amm module factories created.
+     * @param implementation The address of the CoW AMM module implementation.
+     */
+    event CowAmmModuleFactoryCreated(address indexed implementation);
+
+    /**
+     * @notice Index all cow amm modules created.
+     * @param implementation The address of the CoW AMM module implementation.
+     * @param safe The address of the safe that the module will be attached to.
+     */
+    event CowAmmModuleCreated(address indexed implementation, address indexed safe);
+
+    /**
+     * @param _implementation The address of the CoW AMM module implementation.
+     * @param _settler The address of the GPv2Settlement contract.
+     * @param _extensibleFallbackHandler The address of the `ExtensibleFallbackHandler`.
+     * @param _composableCow The address of `ComposableCoW`.
+     * @param _ammHandler The address of the AMM implementation (e.g. `ConstantProduct`) for creating new CoW AMMs.
+     */
+    constructor(
+        CowAmmModule _implementation,
+        address _settler,
+        address _extensibleFallbackHandler,
+        address _composableCow,
+        address _ammHandler
+    ) {
+        // Make sure that the addresses within the implementation correspond correctly to the given addresses
+        bool isSettler = address(_implementation.SETTLER()) == _settler;
+        bool isExtensibleFallbackHandler =
+            address(_implementation.EXTENSIBLE_FALLBACK_HANDLER()) == _extensibleFallbackHandler;
+        bool isComposableCow = address(_implementation.COMPOSABLE_COW()) == _composableCow;
+        bool isAmmHandler = address(_implementation.HANDLER()) == _ammHandler;
+        if (!isSettler || !isExtensibleFallbackHandler || !isComposableCow || !isAmmHandler) {
+            revert InvalidParameter();
+        }
+
+        // Set the implementation
+        IMPLEMENTATION = _implementation;
+
+        emit CowAmmModuleFactoryCreated(address(_implementation));
+    }
+
+    /**
+     * @notice Creates a new CoW AMM module for the given safe.
+     * @param safe The address of the safe that the module will be attached to.
+     * @return The address of the new CoW AMM module.
+     */
+    function create(address safe) external returns (address) {
+        CowAmmModule module = CowAmmModule(address(IMPLEMENTATION).cloneDeterministic(bytes20(safe)));
```

Comment:
> Tricky use of proxies! It's important to note that anyone can create a module at the expected address without calling `create`. What's worse, after that it could be initialized to any other safe, so even if you compute an address with `predictAddress`, you can't be sure it will be safe to use that as a module unless you are certain that you called `create` in the same transaction as you added the module. This can be fixed by having the proxy check its own address as part of `initialize` to guarantee there's no mismatch between salt and `safe`. Then the only worry left is that people could frontrun a call to `create` and cause it to revert by deploying an address in advance, but this is a smaller concern.

`cloneDeterministic` [makes use of `CREATE2`](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/fd81a96f01cc42ef1c9a5399364968d0e07e9e90/contracts/proxy/Clones.sol#L53). As such, `CREATE2` yields a deployment addresses following the [specification](https://eips.ethereum.org/EIPS/eip-1014#specification):

`keccak256( 0xff ++ address ++ salt ++ keccak256(init_code))[12:]`

In the above, `address` corresponds to the deployer (in this case the `CowAmmModuleFactory`).

Therefore, the only way to deploy to an address that corresponds to `predictAddress` is by calling the `create` function on the `CowAmmModuleFactory`. It also ensures that the proxy that is generated, it's `safe` is set to the corresponding address, making it deterministically bound.

There may however be an argument to change the function from `create` here to something that is less likely to get conflated with `CREATE`.

> More in general, I don't think we need to use proxies at all, getting rid of them would reduce the complexity of the setup. An idea would be to drop the `safe` state variable (using `msg.sender` instead) and replace `activeOrder` with a mapping from safes to active orders. Then, I believe everyone could just enable the module at a fixed address and not bother with any call to `create` in advance. I'd also expect it to be more gas efficient or at least comparable: we're trading an `SLOAD` for a `DELEGATECALL`.

This is a great point. It also saves contract spam. I'll drop the proxy in lieu of this approach.

Reply:
> Therefore, the only way to deploy to an address that corresponds to predictAddress is by calling the create function on the CowAmmModuleFactory.

Right, I'm too used to the usual deterministic deployer or any other unchecked deployer that I forgot the deployer address is hashed in as well.

Anyway, the latest changes look like an improvement on the proxy setup!